File: harness/codegen/control_flow/executable.py
from typing import Iterable, Optional
from harness.core import Process
from harness.control_flow import ControlFlowMutex
from harness.codegen.control_flow.base import HarnessControlFlowBaseCodegen, IntOrStrIterable

class HarnessControlFlowExecutableCodegen(HarnessControlFlowBaseCodegen):
    def _prologue(self):
        yield '#include <stdlib.h>'
        yield '#include <stdio.h>'
        yield '#include <stdbool.h>'
        yield '#include <pthread.h>'

    def _declare_mutex(self, mutex: str):
        yield f'static pthread_mutex_t {mutex};'

    def _declare_init_barrier(self, _: Iterable[Process]):
        yield f'static pthread_barrier_t harness_init_barrier;'

    def _open_process_definition(self, process: Process):
        yield f'void *process_{process.mnemonic}(void *arg) {{'
        yield 1
        yield '(void) arg; // Unused'

    def _close_process_definition(self, process):
        yield 'return NULL;'
        yield -1
        yield '}'

    def _open_main_definition(self):
        yield 'int main() {'
        yield 1

    def _close_main_definition(self):
        yield 'return 0;'
        yield -1
        yield '}'

    def _initialize_mutex(self, mutex: str):
        yield f'pthread_mutex_init(&{mutex}, NULL);'

    def _setup_init_barrier(self, processes: Iterable[Process]):
        yield f'pthread_barrier_init(&harness_init_barrier, NULL, {len(list(processes))});'

    def _declare_process(self, process: Process):
        yield f'pthread_t process_thread_{process.mnemonic};'

    def _start_process(self, process: Process):
        yield f'pthread_create(&process_thread_{process.mnemonic}, NULL, process_{process.mnemonic}, NULL);'

    def _join_process(self, process: Process):
        yield f'pthread_join(process_thread_{process.mnemonic}, NULL);'

    def _lock_mutex(self, mutex: str):
        yield f'pthread_mutex_lock(&{mutex});'

    def _unlock_mutex(self, mutex: str):
        yield f'pthread_mutex_unlock(&{mutex});'

    def _init_barrier_wait(self, process: Process, other_processes: Iterable[Process]) -> IntOrStrIterable:
        yield f'pthread_barrier_wait(&harness_init_barrier);'

    def _mutex_set_transition(self, lock: Iterable[ControlFlowMutex], unlock: Iterable[ControlFlowMutex], rollback_on_failure: Optional[str]) -> IntOrStrIterable:
        lock_mtx = sorted(
            (
                mtx
                for mtx in lock
            ), key=lambda mtx: mtx.identifier
        )
        unlock_mtx = reversed(sorted(
            (
                mtx
                for mtx in unlock
            ), key=lambda mtx: mtx.identifier
        ))
        
        if lock_mtx:
            if rollback_on_failure is not None:
                yield 'for (;;) {'
                yield 1

                for mtx_index, mtx in enumerate(lock_mtx):
                    yield f'if (pthread_mutex_trylock(&{self._mutex_name(mtx)})) {{'
                    yield 1

                    for i in range(mtx_index - 1, -1, -1):
                        yield f'pthread_mutex_unlock(&{self._mutex_name(lock_mtx[i])});'
                    if rollback_on_failure is not None:
                        yield f'goto {rollback_on_failure};'
                    else:
                        yield 'continue;'

                    yield -1
                    yield '}'
                yield 'break;'

                yield -1
                yield '}'
            else:
                for mtx in lock_mtx:
                    yield from self._lock_mutex(self._mutex_name(mtx))
        for mtx in unlock_mtx:
            yield from self._unlock_mutex(self._mutex_name(mtx))

    def _random(self, max: int) -> str:
        return f'(rand() % {max})'

File: harness2/src/harness/codegen/executable.rs
use crate::harness::{control_flow::mutex::ControlFlowMutexID, core::{error::HarnessError, process::ProcessID}};

use super::base::{CodegenOutput, ControlFlowCodegen};

pub struct ControlFlowExecutableCodegen;

impl ControlFlowExecutableCodegen {
    pub fn new() -> ControlFlowExecutableCodegen {
        ControlFlowExecutableCodegen {}
    }
}

impl<Output: CodegenOutput> ControlFlowCodegen<Output> for ControlFlowExecutableCodegen {
    fn generate_prologue(&self, output: &mut Output) -> Result<(), HarnessError> {
        output.write_line("#include <stdlib.h>")?;
        output.write_line("#include <stdio.h>")?;
        output.write_line("#include <stdbool.h>")?;
        output.write_line("#include <pthread.h>")?;
        Ok(())
    }
    
    fn declare_mutex(&self, output: &mut Output, mutex: ControlFlowMutexID) -> Result<(), HarnessError> {
        output.write_line(format!("static pthread_mutex_t mutex{};", Into::<u64>::into(mutex)))?;
        Ok(())
    }

    fn declare_init_barrier(&self, output: &mut Output, _: impl Iterator<Item = ProcessID>) -> Result<(), HarnessError> {
        output.write_line("static pthread_barrier_t harness_init_barrier;")?;
        Ok(())
    }

    fn begin_process_definition(&self, output: &mut Output, process: ProcessID) -> Result<(), HarnessError> {
        output.write_line(format!("void *process{}(void *arg) {{", Into::<u64>::into(process)))?;
        output.indent_up();
        output.write_line("(void) arg; // Unused")?;
        Ok(())
    }

    fn end_process_definition(&self, output: &mut Output, _: ProcessID) -> Result<(), HarnessError> {
        output.write_line("return NULL;")?;
        output.indent_down();
        output.write_line("}")?;
        Ok(())
    }

    fn begin_main_definition(&self, output: &mut Output) -> Result<(), HarnessError> {
        output.write_line("int main() {")?;
        output.indent_up();
        Ok(())
    }

    fn end_main_definition(&self, output: &mut Output) -> Result<(), HarnessError> {
        output.write_line("return 0;")?;
        output.indent_down();
        output.write_line("}")?;
        Ok(())
    }

    fn setup_init_barrier(&self, output: &mut Output, processes: impl Iterator<Item = ProcessID>) -> Result<(), HarnessError> {
        output.write_line(format!("pthread_barrier_init(&harness_init_barrier, NULL, {});",
            processes.count()))?;
        Ok(())
    }

    fn initialize_mutex(&self, output: &mut Output, mutex: ControlFlowMutexID) -> Result<(), HarnessError> {
        output.write_line(format!("pthread_mutex_init(&mutex{}, NULL);",
            Into::<u64>::into(mutex)))?;
        Ok(())
    }

    fn declare_process_thread(&self, output: &mut Output, process: ProcessID) -> Result<(), HarnessError> {
        output.write_line(format!("pthread_t process_thread{};",
            Into::<u64>::into(process)))?;
        Ok(())
    }

    fn start_process_thread(&self, output: &mut Output, process: ProcessID) -> Result<(), HarnessError> {
        output.write_line(format!("pthread_create(&process_thread{}, NULL, process{}, NULL);",
            Into::<u64>::into(process), Into::<u64>::into(process)))?;
        Ok(())
    }

    fn join_process_thread(&self, output: &mut Output, process: ProcessID) -> Result<(), HarnessError> {
        output.write_line(format!("pthread_join(process_thread{}, NULL);",
            Into::<u64>::into(process)))?;
        Ok(())
    }

    fn lock_mutex(&self, output: &mut Output, mutex: ControlFlowMutexID) -> Result<(), HarnessError> {
        output.write_line(format!("pthread_mutex_lock(&mutex{});",
            Into::<u64>::into(mutex)))?;
        Ok(())
    }

    fn unlock_mutex(&self, output: &mut Output, mutex: ControlFlowMutexID) -> Result<(), HarnessError> {
        output.write_line(format!("pthread_mutex_unlock(&mutex{});",
            Into::<u64>::into(mutex)))?;
        Ok(())
    }

    fn do_synchronization(&self, output: &mut Output, lock: impl Iterator<Item = ControlFlowMutexID>, unlock: impl Iterator<Item = ControlFlowMutexID>, rollback_on_failure: Option<&str>) -> Result<(), HarnessError> {
        if let Some(rollback_label) = rollback_on_failure {
            output.write_line("for (;;) {")?;
            output.indent_up();

            let lock = lock.collect::<Vec<_>>();
            for (index, &mutex) in lock.iter().enumerate() {
                output.write_line(format!("if (pthread_mutex_trylock(&mutex{})) {{",
                    Into::<u64>::into(mutex)))?;
                output.indent_up();

                for i in (0..index).rev() {
                    output.write_line(format!("pthread_mutex_unlock(&mutex{});",
                        Into::<u64>::into(*lock.get(i).unwrap())))?;
                }
                output.write_line(format!("goto {};", rollback_label))?;

                output.indent_down();
                output.write_line("}")?;
            }

            output.write_line("break;")?;
            output.indent_down();
            output.write_line("}")?;
        } else {
            for mutex in lock {
                self.lock_mutex(output, mutex)?;
            }
        }
        for mutex in unlock {
            self.unlock_mutex(output, mutex)?
        }
        Ok(())
    }

    fn wait_init_barrier(&self, output: &mut Output, _: ProcessID, _: impl Iterator<Item = ProcessID>) -> Result<(), HarnessError> {
        output.write_line("pthread_barrier_wait(&harness_init_barrier);")?;
        Ok(())
    }

    fn generate_random(&self, max: u32) -> String {
        format!("rand() % {}", max)
    }
}
